File: packages/htf-js/htf/framework.js
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025-2026 ZHU Shengling
(function (global) {
  "use strict";

  const HTF = global.HTF || (global.HTF = {});

  class HTFFramework {
    constructor({ timeframes, coordinator } = {}) {
      this.timeframes = timeframes || {};
      this.coordinator = coordinator;
      this.last_output = {};
    }

    reset() {
      const values = this.timeframes instanceof Map ? Array.from(this.timeframes.values()) : Object.values(this.timeframes);
      values.forEach((tf) => {
        if (tf && typeof tf.reset === "function") {
          tf.reset();
        }
      });
      this.last_output = {};
    }

    on_new_record(record) {
      const entries = this.timeframes instanceof Map ? Array.from(this.timeframes.entries()) : Object.entries(this.timeframes);
      entries.forEach(([, tf]) => {
        if (tf && typeof tf.on_new_record === "function") {
          tf.on_new_record(record);
        }
      });

      const states = {};
      entries.forEach(([name, tf]) => {
        states[name] = new HTF.TimeframeState({
          name: tf.name,
          role: tf.role,
          features: tf.features,
          signal: tf.signal,
        });
      });

      const coord = this.coordinator ? this.coordinator.update(states, record) : {};
      this.last_output = { states, coordination: coord };
      return this.last_output;
    }
  }

  HTF.HTFFramework = HTFFramework;
})(typeof window !== "undefined" ? window : globalThis);

File: packages/htf-py/htf/framework.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2025-2026 ZHU Shengling
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Any, Dict

from .coordinator import MultiScaleCoordinator, TimeframeState
from .timeframe import TimeframeView


@dataclass
class HTFFramework:
    timeframes: Dict[str, TimeframeView]
    coordinator: MultiScaleCoordinator

    last_output: Dict[str, Any] = field(default_factory=dict, init=False)

    def reset(self) -> None:
        for tf in self.timeframes.values():
            tf.reset()
        self.last_output = {}

    def on_new_record(self, record: Mapping[str, Any]) -> Dict[str, Any]:
        for tf in self.timeframes.values():
            tf.on_new_record(record)

        states: Dict[str, TimeframeState] = {
            name: TimeframeState(
                name=tf.name,
                role=tf.role,
                features=tf.features,
                signal=tf.signal,
            )
            for name, tf in self.timeframes.items()
        }

        coord = self.coordinator.update(states, record)
        self.last_output = {"states": states, "coordination": coord}
        return self.last_output
